GKE Ingestion Jobs Failing

## Expected Behavior 
Ingestion jobs successfully run and bigquery table (`feast.customer_project_customer_transactions_v1`) gets populated with data.

## Current Behavior
```
>>> client.ingest("customer_transactions", customer_features)
Waiting for feature set to be ready for ingestion...
  0%|                                                                                                | 0/155 [04:48<?, ?rows/s]
Ingestion complete!

Ingestion statistics:
Success: 0/155
Removing temporary file(s)...
```

BQ table is empty:
```
SELECT COUNT(*) FROM `feast.customer_project_customer_transactions_v1`;
```
is 0.

## Steps to reproduce
Follow latest [GKE setup docs](https://docs.feast.dev/getting-started/installing-feast#google-kubernetes-engine) with the following differences:

* `my-project` as project ID
* Use image 0.4.2
* Use `NodePort` to expose service to local client
* In basic example, use `$FEAST_CORE_URL` and `$FEAST_BATCH_SERVING_URL`
* Open up nodeports in GCP firewall 32090, 32091, 32092:

```
gcloud compute firewall-rules create feast-core-port --allow tcp:32090
gcloud compute firewall-rules create feast-online-port --allow tcp:32091
gcloud compute firewall-rules create feast-batch-port --allow tcp:32092
```

`my-feast-values.yaml` ends up looking like:

```
feast-core:
  enabled: true
  image:
    tag: "0.4.2"
  jvmOptions:
  - -Xms1024m
  - -Xmx1024m
  resources:
    requests:
      cpu: 1000m
      memory: 1024Mi
  service:
    type: NodePort
    grpc:
      nodePort: 32090
  gcpServiceAccount:
    useExistingSecret: true
feast-serving-online:
  enabled: true
  redis:
    enabled: true
  image:
    tag: "0.4.2"
  jvmOptions:
  - -Xms1024m
  - -Xmx1024m
  resources:
    requests:
      cpu: 500m
      memory: 1024Mi
  service:
    type: NodePort
    grpc:
      nodePort: 32091
  store.yaml:
    name: redis
    type: REDIS
    redis_config:
      port: 6379
    subscriptions:
    - name: "*"
      project: "*"
      version: "*"
feast-serving-batch:
  enabled: true
  redis:
    enabled: false
  image:
    tag: "0.4.2"
  jvmOptions:
  - -Xms1024m
  - -Xmx1024m
  resources:
    requests:
      cpu: 500m
      memory: 1024Mi
  service:
    type: NodePort
    grpc:
      nodePort: 32092
  gcpServiceAccount:
    useExistingSecret: true
  application.yaml:
    feast:
      jobs:
        staging-location: gs://my-project_feast_bucket/serving/batch
        store-type: REDIS
        store-options:
          host: localhost
          port: 6379
  store.yaml:
    name: bigquery
    type: BIGQUERY
    bigquery_config:
      project_id: my-project
      dataset_id: feast
    subscriptions:
    - name: "*"
      project: "*"
      version: "*"
```

### Specifications

- Version: 0.4.2, latest master python SDK (2a33f7bbbe6bd92f292a128102f864cd67e95660)
- Platform: Installing GKE on local Mac OSX
- Subsystem: Python 3.7.6, helm v2.16.1, kubectl v1.17.0

## Possible Solution

Default `Kafka` configs might need adjusting? Maybe related to `NodePort` config?

`kubectl` logs for `feast-feast-core` appear to show kafka jobs were successfully created and sent to `DirectRunner`:

```
22:57:52 [pool-5-thread-1] INFO  feast.ingestion.ImportJob - Starting import job with settings:
Current Settings:
  appName: DirectRunnerJobManager
  blockOnRun: false
  enforceEncodability: true
  enforceImmutability: true
  featureSetJson: [{
  "name": "customer_transactions",
  "version": 1,
  "entities": [{
    "name": "customer_id",
    "valueType": "INT64"
  }],
  "features": [{
    "name": "total_transactions",
    "valueType": "INT64"
  }, {
    "name": "daily_transactions",
    "valueType": "DOUBLE"
  }],
  "maxAge": "86400s",
  "source": {
    "type": "KAFKA",
    "kafkaSourceConfig": {
      "bootstrapServers": "feast-kafka:9092",
      "topic": "feast"
    }
  },
  "project": "customer_project"
}]
  gcsPerformanceMetrics: false
  optionsId: 0
  project:
  runner: class org.apache.beam.runners.direct.DirectRunner
  stableUniqueNames: WARNING
  storeJson: [{
  "name": "bigquery",
  "type": "BIGQUERY",
  "subscriptions": [{
    "name": "*",
    "version": "*",
    "project": "*"
  }],
  "bigqueryConfig": {
    "projectId": "my-project",
    "datasetId": "feast"
  }
}]

22:57:54 [pool-5-thread-1] INFO  feast.ingestion.utils.StoreUtil - Writing to existing BigQuery table 'my-project:feast.customer_project_customer_transactions_v1'
22:57:54 [pool-6-thread-1] INFO  org.apache.beam.sdk.io.kafka.KafkaUnboundedSource - Partitions assigned to split 0 (total 1): feast-0
2020-01-09 22:57:54.984 AUDIT feast-feast-core-dc485b44d-qg75w --- [pool-6-thread-1] f.c.l.AuditLogger                        : {action=STATUS_CHANGE, detail=Job submitted to runner DirectRunner with ext id kafka-to-redis1578610671583., id=kafka-to-redis1578610671583, resource=JOB, timestamp=Thu Jan 09 22:57:54 UTC 2020}
22:57:55 [pool-5-thread-1] INFO  org.apache.beam.sdk.io.kafka.KafkaUnboundedSource - Partitions assigned to split 0 (total 1): feast-0
2020-01-09 22:57:55.373 AUDIT feast-feast-core-dc485b44d-qg75w --- [pool-5-thread-1] f.c.l.AuditLogger                        : {action=STATUS_CHANGE, detail=Job submitted to runner DirectRunner with ext id kafka-to-bigquery1578610671583., id=kafka-to-bigquery1578610671583, resource=JOB, timestamp=Thu Jan 09 22:57:55 UTC 2020}
22:57:55 [pool-2-thread-1] INFO  feast.core.service.JobCoordinatorService - Updating feature set status
22:57:58 [direct-runner-worker] INFO  org.apache.beam.sdk.io.kafka.KafkaUnboundedSource - Reader-0: reading from feast-0 starting at offset 0
22:57:58 [direct-runner-worker] INFO  org.apache.beam.sdk.io.kafka.KafkaUnboundedSource - Reader-0: reading from feast-0 starting at offset 0
```

And a `feast` topic was successfully created according to the `kafka-config` pod:
```
Waiting for Zookeeper...
Waiting for Kafka...
Applying runtime configuration using confluentinc/cp-kafka:5.0.1
Created topic "feast".
Configs for topic 'feast' are 
```

But, I'm getting errors thrown related to topics in `feast-kafka-0` logs:
```
[2020-01-09 22:57:57,799] INFO [Log partition=__consumer_offsets-2, dir=/opt/kafka/data/logs] Truncating to 0 has no effect as the largest offset in the log is -1 (kafka.log.Log)
[2020-01-09 22:57:57,804] ERROR [ReplicaFetcher replicaId=0, leaderId=1, fetcherId=0] Error for partition __consumer_offsets-8 at offset 0 (kafka.server.ReplicaFetcherThread)
org.apache.kafka.common.errors.UnknownTopicOrPartitionException: This server does not host this topic-partition.
[2020-01-09 22:57:57,804] ERROR [ReplicaFetcher replicaId=0, leaderId=1, fetcherId=0] Error for partition __consumer_offsets-35 at offset 0 (kafka.server.ReplicaFetcherThread)
org.apache.kafka.common.errors.UnknownTopicOrPartitionException: This server does not host this topic-partition.
[2020-01-09 22:57:57,805] ERROR [ReplicaFetcher replicaId=0, leaderId=1, fetcherId=0] Error for partition __consumer_offsets-41 at offset 0 (kafka.server.ReplicaFetcherThread)
```

Happy to share any other relevant logs. I'm admittedly not familiar with Kafka, so I could be off here. Just trying to get the GKE feast guide + basic example working end to end. Once it works, happy to put up a PR to update the guide for 0.4.X.

Okay, getting close.

It was related to the wrong IP. After updating my bootstrapServers in my helm chart, the existing `customer_fs` was using my old config, so it was using `feast-kafka:9092` as the broker. I'm not sure if that's expected behavior?

I forced it to use new config by creating a new feature set and was able to ingest successfully!

```
100%|█████████████████████████████████████████████████████████████████████████████████████| 155/155 [00:00<00:00, 227.54rows/s]
Ingestion complete!███████████████████████████████████████████████████████████████████████| 155/155 [00:00<00:00, 227.78rows/s]
```

Batch serving is throwing me this error though, but hopefully a similar fix:
```
2020-01-16 23:57:49.861 ERROR feast-feast-serving-batch-757cd5d649-4xj8d --- [ault-executor-1] f.s.s.RedisBackedJobService              : Failed to upsert job: Failed connecting to host localhost:6379
2020-01-16 23:57:49.955  WARN feast-feast-serving-batch-757cd5d649-4xj8d --- [ault-executor-1] f.s.c.ServingServiceGRpcController       : Failed to get Job
redis.clients.jedis.exceptions.JedisConnectionException: Failed connecting to host localhost:6379
	at redis.clients.jedis.Connection.connect(Connection.java:207)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:101)
	at redis.clients.jedis.Connection.sendCommand(Connection.java:126)
	at redis.clients.jedis.Connection.sendCommand(Connection.java:117)
	at redis.clients.jedis.Jedis.get(Jedis.java:155)
	at feast.serving.service.RedisBackedJobService.get(RedisBackedJobService.java:45)
	at feast.serving.service.BigQueryServingService.getJob(BigQueryServingService.java:171)
	at feast.serving.controller.ServingServiceGRpcController.getJob(ServingServiceGRpcController.java:100)
	at feast.serving.ServingServiceGrpc$MethodHandlers.invoke(ServingServiceGrpc.java:503)
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:283)
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:710)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at redis.clients.jedis.Connection.connect(Connection.java:184)
	... 16 more
2020-01-16 23:58:00.336 ERROR feast-feast-serving-batch-757cd5d649-4xj8d --- [       Thread-6] f.s.s.RedisBackedJobService              : Failed to upsert job: Failed connecting to host localhost:6379
2020-01-16 23:58:49.026  INFO feast-feast-serving-batch-757cd5d649-4xj8d --- [pool-3-thread-1] f.s.s.CachedSpecService                  : loaded store config at /etc/feast/feast-serving/store.yaml: 
```

> Okay, getting close.
> 
> It was related to the wrong IP. After updating my bootstrapServers in my helm chart, the existing `customer_fs` was using my old config, so it was using `feast-kafka:9092` as the broker. I'm not sure if that's expected behavior?
> 
> I forced it to use new config by creating a new feature set and was able to ingest successfully!
> 
> ```
> 100%|█████████████████████████████████████████████████████████████████████████████████████| 155/155 [00:00<00:00, 227.54rows/s]
> Ingestion complete!███████████████████████████████████████████████████████████████████████| 155/155 [00:00<00:00, 227.78rows/s]
> ```
> 
> Batch serving is throwing me this error though, but hopefully a similar fix:
> 
> ```
> 2020-01-16 23:57:49.861 ERROR feast-feast-serving-batch-757cd5d649-4xj8d --- [ault-executor-1] f.s.s.RedisBackedJobService              : Failed to upsert job: Failed connecting to host localhost:6379
> 2020-01-16 23:57:49.955  WARN feast-feast-serving-batch-757cd5d649-4xj8d --- [ault-executor-1] f.s.c.ServingServiceGRpcController       : Failed to get Job
> redis.clients.jedis.exceptions.JedisConnectionException: Failed connecting to host localhost:6379
> 	at redis.clients.jedis.Connection.connect(Connection.java:207)
> 	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:101)
> 	at redis.clients.jedis.Connection.sendCommand(Connection.java:126)
> 	at redis.clients.jedis.Connection.sendCommand(Connection.java:117)
> 	at redis.clients.jedis.Jedis.get(Jedis.java:155)
> 	at feast.serving.service.RedisBackedJobService.get(RedisBackedJobService.java:45)
> 	at feast.serving.service.BigQueryServingService.getJob(BigQueryServingService.java:171)
> 	at feast.serving.controller.ServingServiceGRpcController.getJob(ServingServiceGRpcController.java:100)
> 	at feast.serving.ServingServiceGrpc$MethodHandlers.invoke(ServingServiceGrpc.java:503)
> 	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
> 	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:283)
> 	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:710)
> 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
> 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: java.net.ConnectException: Connection refused (Connection refused)
> 	at java.net.PlainSocketImpl.socketConnect(Native Method)
> 	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
> 	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
> 	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
> 	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
> 	at java.net.Socket.connect(Socket.java:589)
> 	at redis.clients.jedis.Connection.connect(Connection.java:184)
> 	... 16 more
> 2020-01-16 23:58:00.336 ERROR feast-feast-serving-batch-757cd5d649-4xj8d --- [       Thread-6] f.s.s.RedisBackedJobService              : Failed to upsert job: Failed connecting to host localhost:6379
> 2020-01-16 23:58:49.026  INFO feast-feast-serving-batch-757cd5d649-4xj8d --- [pool-3-thread-1] f.s.s.CachedSpecService                  : loaded store config at /etc/feast/feast-serving/store.yaml: 
> ```

I have also been going through the helm installation and resolved most issues, but I have discovered new holes that I am trying to plug. Here is my current list of changes that I want to make to the docs/charts

1. Open up firewall to allow NodePorts
```
gcloud compute firewall-rules create feast-core-port --allow tcp:32090
gcloud compute firewall-rules create feast-online-port --allow tcp:32091
gcloud compute firewall-rules create feast-batch-port --allow tcp:32092
gcloud compute firewall-rules create feast-redis-port --allow tcp:32101
gcloud compute firewall-rules create feast-kafka --allow tcp:31090-31095
```

2. Create a bucket and update the helm values in batch serving to have the correct path.
3. Update the `dataset` and `project` in `store.yaml` key in helm values
4. Update `store-options:` to use an external IP and NodePort. Feast batch serving uses Redis as its job store to retain state, so it needs to be able to connect to Redis. Its better to use the Redis that online serving uses so that we don't need to deploy two servers. (FYI: this is tech debt and will be cleared in the future).
```
application.yaml:
    feast:
      jobs: 
        staging-location: gs://[bucket]/staging
        store-type: REDIS
        store-options:
          host: [EXTERNAL IP]
          port: 32101
```

```
update 
    master:
      service:
        nodePort: 36379
        type: NodePort
```
5. Update image tag to `0.4.3`
6. Add NodePort to all services, for example
```
  service:
    type: NodePort
    grpc:
      nodePort: 32092
```
7. Set the Feast Online Serving store to use an external IP and port
```
 store.yaml:
    name: redis
    type: REDIS
    redis_config:
      host: [EXTERNAL IP]
      port: 32101
```
- Make sure that a Redis instance is deployed with online serving and that it uses NodePort
```
  redis:
    enabled: true
    master:
      service:
        nodePort: 32101
        type: NodePort
```
8. Make sure Kafka is also using NodePort
```
  kafka:
    enabled: true
    persistence:
      enabled: false
    external:
      enabled: true
      type: NodePort
      domain: [EXTERNAL IP]
    configurationOverrides:
      "advertised.listeners": |-
        EXTERNAL://[EXTERNAL IP]:$((31090 + ${KAFKA_BROKER_ID}))
      "listener.security.protocol.map": |-
        PLAINTEXT:PLAINTEXT,EXTERNAL:PLAINTEXT
```

I think the problem you have right now is that Feast Batch Serving is trying to find that Redis instance but it cant find it on localhost. If you are using `DirectRunner` then the Redis does not need to be exposed outside of the cluster and you can just provide Feast Batch Serving with the internal IP and port of the Redis service. If you are going to use Dataflow then you need to use the externally exposed Redis IP (as I have done above) when configuring Feast Batch Serving.

--- a/docs/getting-started/installing-feast.md
+++ b/docs/getting-started/installing-feast.md
@@ -268,7 +268,7 @@ bq mk ${FEAST_BIGQUERY_DATASET_ID}
 Create the service account that Feast will run as:
 
 ```bash
-gcloud iam service-accounts create ${FEAST_SERVICE_ACCOUNT_NAME}
+gcloud iam service-accounts create ${FEAST_S_ACCOUNT_NAME}
 
 gcloud projects add-iam-policy-binding ${FEAST_GCP_PROJECT_ID} \
   --member serviceAccount:${FEAST_S_ACCOUNT_NAME}@${FEAST_GCP_PROJECT_ID}.iam.gserviceaccount.com \
@@ -324,6 +324,15 @@ PING 10.123.114.11 (10.203.164.22) 56(84) bytes of data.
 64 bytes from 10.123.114.11: icmp_seq=2 ttl=63 time=51.2 ms
 ```
 
+Add firewall rules in gcloud to open up ports:
+```bash
+gcloud compute firewall-rules create feast-core-port --allow tcp:32090
+gcloud compute firewall-rules create feast-online-port --allow tcp:32091
+gcloud compute firewall-rules create feast-batch-port --allow tcp:32092
+gcloud compute firewall-rules create feast-redis-port --allow tcp:32101
+gcloud compute firewall-rules create feast-kafka-ports --allow tcp:31090-31095
+```
+
 ### 3. Set up Helm
 
 Run the following command to provide Tiller with authorization to install Feast:
@@ -377,7 +386,8 @@ cp values.yaml my-feast-values.yaml
 Update `my-feast-values.yaml` based on your GCP and GKE environment.
 
 * Required fields are paired with comments which indicate whether they need to be replaced.
-* All occurrences of `feast.example.com` should be replaced with either your domain name or the IP stored in `$FEAST_IP`.
+* All occurrences of `EXTERNAL_IP` should be replaced with either your domain name or the IP stored in `$FEAST_IP`.
+* Replace all occurrences of `YOUR_BUCKET_NAME` with your bucket name stored in `$FEAST_GCS_BUCKET`
 
 Install the Feast Helm chart:
 
@@ -421,4 +431,3 @@ feast config set serving_url ${FEAST_ONLINE_SERVING_URL}
 ```
 
 That's it! You can now start to use Feast!
-

--- a/infra/charts/feast/values.yaml
+++ b/infra/charts/feast/values.yaml
@@ -2,29 +2,29 @@
 # - Feast Core
 # - Feast Serving Online
 # - Feast Serving Batch
-# 
+#
 # The configuration for different components can be referenced from:
 # - charts/feast-core/values.yaml
 # - charts/feast-serving/values.yaml
 #
 # Note that "feast-serving-online" and "feast-serving-batch" are
 # aliases to "feast-serving" chart since in typical scenario two instances
-# of Feast Serving: online and batch will be deployed. Both described 
+# of Feast Serving: online and batch will be deployed. Both described
 # using the same chart "feast-serving".
 #
 # The following are default values for typical Feast deployment, but not
 # for production setting. Refer to "values-production.yaml" for recommended
 # values in production environment.
-# 
-# Note that the import job by default uses DirectRunner 
+#
+# Note that the import job by default uses DirectRunner
 # https://beam.apache.org/documentation/runners/direct/
 # in this configuration since it allows Feast to run in more environments
 # (unlike DataflowRunner which requires Google Cloud services).
-# 
-# A secret containing Google Cloud service account JSON key is required 
-# in this configuration. 
+#
+# A secret containing Google Cloud service account JSON key is required
+# in this configuration.
 # https://cloud.google.com/iam/docs/creating-managing-service-accounts
-# 
+#
 # The Google Cloud service account must have the following roles:
 # - bigquery.dataEditor
 # - bigquery.jobUser
@@ -32,12 +32,13 @@
 # Assuming a service account JSON key file has been downloaded to
 # (please name the file key.json):
 # /home/user/key.json
-# 
+#
 # Run the following command to create the secret in your Kubernetes cluster:
 #
 # kubectl create secret generic feast-gcp-service-account \
 #   --from-file=/home/user/key.json
 #
+# Replace every instance of EXTERNAL_IP with the external IP of your GKE cluster
 
 # ============================================================
 # Feast Core
@@ -51,12 +52,15 @@ feast-core:
   # to the client. These instances of Feast Serving however can still use
   # the same shared Feast Core.
   enabled: true
-  # jvmOptions are options that will be passed to the Java Virtual Machine (JVM) 
+  # Specify what image tag to use. Keep this consistent for all components
+  image:
+    tag: "0.4.3"
+  # jvmOptions are options that will be passed to the Java Virtual Machine (JVM)
   # running Feast Core.
   #
   # For example, it is good practice to set min and max heap size in JVM.
   # https://stackoverflow.com/questions/6902135/side-effect-for-increasing-maxpermsize-and-max-heap-size
-  jvmOptions: 
+  jvmOptions:
   - -Xms1024m
   - -Xmx1024m
   # resources that should be allocated to Feast Core.
@@ -68,18 +72,43 @@ feast-core:
       memory: 2048Mi
   # gcpServiceAccount is the Google service account that Feast Core will use.
   gcpServiceAccount:
-    # useExistingSecret specifies Feast to use an existing secret containing 
+    # useExistingSecret specifies Feast to use an existing secret containing
     # Google Cloud service account JSON key file.
-    # 
+    #
     # This is the only supported option for now to use a service account JSON.
     # Feast admin is expected to create this secret before deploying Feast.
     useExistingSecret: true
     existingSecret:
       # name is the secret name of the existing secret for the service account.
-      name: feast-gcp-service-account 
+      name: feast-gcp-service-account
       # key is the secret key of the existing secret for the service account.
       # key is normally derived from the file name of the JSON key file.
       key: key.json
+  # Setting service.type to NodePort exposes feast-core service at a static port
+  service:
+    type: NodePort
+    grpc:
+      # this is the port that is exposed outside of the cluster
+      nodePort: 32090
+  # Make kafka externally accessible using NodePort
+  # Please set EXTERNAL_IP to your cluster's external IP
+  kafka:
+    external:
+      enabled: true
+      type: NodePort
+      domain: EXTERNAL_IP
+    configurationOverrides:
+      "advertised.listeners": |-
+        EXTERNAL://EXTERNAL_IP:$((31090 + ${KAFKA_BROKER_ID}))
+      "listener.security.protocol.map": |-
+        PLAINTEXT:PLAINTEXT,EXTERNAL:PLAINTEXT
+  application.yaml:
+    feast:
+      stream:
+        options:
+          # Point to one of your Kafka brokers
+          # Please set EXTERNAL_IP to your cluster's external IP
+          bootstrapServers: EXTERNAL_IP:31090
 
 # ============================================================
 # Feast Serving Online
@@ -88,14 +117,22 @@ feast-core:
 feast-serving-online:
   # enabled specifies whether to install Feast Serving Online component.
   enabled: true
+  # Specify what image tag to use. Keep this consistent for all components
+  image:
+    tag: "0.4.3"
   # redis.enabled specifies whether Redis should be installed as part of Feast Serving.
-  # 
+  #
   # If enabled is set to "false", Feast admin has to ensure there is an
   # existing Redis running outside Feast, that Feast Serving can connect to.
+  # master.service.type set to NodePort exposes Redis to outside of the cluster
   redis:
     enabled: true
+    master:
+      service:
+        nodePort: 32101
+        type: NodePort
   # jvmOptions are options that will be passed to the Feast Serving JVM.
-  jvmOptions: 
+  jvmOptions:
   - -Xms1024m
   - -Xmx1024m
   # resources that should be allocated to Feast Serving.
@@ -105,23 +142,28 @@ feast-serving-online:
       memory: 1024Mi
     limits:
       memory: 2048Mi
+  # Make service accessible to outside of cluster using NodePort
+  service:
+    type: NodePort
+    grpc:
+      nodePort: 32091
   # store.yaml is the configuration for Feast Store.
-  # 
+  #
   # Refer to this link for more description:
   # https://github.com/gojek/feast/blob/79eb4ab5fa3d37102c1dca9968162a98690526ba/protos/feast/core/Store.proto
   store.yaml:
     name: redis
     type: REDIS
     redis_config:
-      # If redis.enabled is set to false, Feast admin should uncomment and 
-      # set the host value to an "existing" Redis instance Feast will use as 
-      # online Store. 
-      # 
-      # Else, if redis.enabled is set to true, no additional configuration is
-      # required.
+      # If redis.enabled is set to false, Feast admin should uncomment and
+      # set the host value to an "existing" Redis instance Feast will use as
+      # online Store. Also use the correct port for that existing instance.
       #
+      # Else, if redis.enabled is set to true, replace EXTERNAL_IP with your
+      # cluster's external IP.
       # host: redis-host
-      port: 6379
+      host: EXTERNAL_IP
+      port: 32101
     subscriptions:
     - name: "*"
       project: "*"
@@ -134,14 +176,17 @@ feast-serving-online:
 feast-serving-batch:
   # enabled specifies whether to install Feast Serving Batch component.
   enabled: true
+  # Specify what image tag to use. Keep this consistent for all components
+  image:
+    tag: "0.4.3"
   # redis.enabled specifies whether Redis should be installed as part of Feast Serving.
-  # 
+  #
   # This is usually set to "false" for Feast Serving Batch because the default
   # store is BigQuery.
   redis:
     enabled: false
   # jvmOptions are options that will be passed to the Feast Serving JVM.
-  jvmOptions: 
+  jvmOptions:
   - -Xms1024m
   - -Xmx1024m
   # resources that should be allocated to Feast Serving.
@@ -151,17 +196,22 @@ feast-serving-batch:
       memory: 1024Mi
     limits:
       memory: 2048Mi
+  # Make service accessible to outside of cluster using NodePort
+  service:
+    type: NodePort
+    grpc:
+      nodePort: 32092
   # gcpServiceAccount is the service account that Feast Serving will use.
   gcpServiceAccount:
-    # useExistingSecret specifies Feast to use an existing secret containing 
+    # useExistingSecret specifies Feast to use an existing secret containing
     # Google Cloud service account JSON key file.
-    # 
+    #
     # This is the only supported option for now to use a service account JSON.
     # Feast admin is expected to create this secret before deploying Feast.
     useExistingSecret: true
     existingSecret:
       # name is the secret name of the existing secret for the service account.
-      name: feast-gcp-service-account 
+      name: feast-gcp-service-account
       # key is the secret key of the existing secret for the service account.
       # key is normally derived from the file name of the JSON key file.
       key: key.json
@@ -172,28 +222,33 @@ feast-serving-batch:
   # for a complete list and description of the configuration.
   application.yaml:
     feast:
-      jobs: 
-        # staging-location specifies the URI to store intermediate files for 
+      jobs:
+        # staging-location specifies the URI to store intermediate files for
         # batch serving (required if using BigQuery as Store).
-        # 
-        # Please set the value to an "existing" Google Cloud Storage URI that 
-        # Feast serving has write access to.
-        staging-location: gs://bucket/path
-        # Type of store to store job metadata. 
         #
-        # This default configuration assumes that Feast Serving Online is 
+        # Please set the value to an "existing" Google Cloud Storage URI that
+        # Feast serving has write access to.
+        staging-location: gs://YOUR_BUCKET_NAME/serving/batch
+        # Type of store to store job metadata.
+        #
+        # This default configuration assumes that Feast Serving Online is
         # enabled as well. So Feast Serving Batch will share the same
         # Redis instance to store job statuses.
         store-type: REDIS
+        store-options:
+          # Use the externally exposed redis instance deployed by Online service
+          # Please set EXTERNAL_IP to your cluster's external IP
+          host: EXTERNAL_IP
+          port: 32101
   # store.yaml is the configuration for Feast Store.
-  # 
+  #
   # Refer to this link for more description:
   # https://github.com/gojek/feast/blob/79eb4ab5fa3d37102c1dca9968162a98690526ba/protos/feast/core/Store.proto
   store.yaml:
     name: bigquery
     type: BIGQUERY
     bigquery_config:
-      # project_id specifies the Google Cloud Project. Please set this to the 
+      # project_id specifies the Google Cloud Project. Please set this to the
       # project id you are using BigQuery in.
       project_id: PROJECT_ID
       # dataset_id specifies an "existing" BigQuery dataset Feast Serving Batch

